File: backend/app/schemas/adaptation.py
"""改编模块 Pydantic 出入参 schema。"""
from datetime import datetime
from typing import Any, List, Literal, Optional

from pydantic import BaseModel, Field

EntityType = Literal["person", "place", "prop", "era_term", "other"]


class AdaptationProjectCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    raw_text: Optional[str] = None
    intent: Optional[str] = None
    intensity: int = Field(default=2, ge=1, le=3)
    era_target: Optional[str] = None


class AdaptationProjectUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=200)
    intent: Optional[str] = None
    intensity: Optional[int] = Field(default=None, ge=1, le=3)
    era_target: Optional[str] = None


class MappingEntryIn(BaseModel):
    entity_type: EntityType
    original_text: str = Field(min_length=1, max_length=200)
    replacement_text: Optional[str] = Field(default=None, max_length=200)
    locked: bool = False
    notes: Optional[str] = None
    order_index: int = 0


class MappingEntryOut(MappingEntryIn):
    id: int


class MappingsBulkPut(BaseModel):
    entries: List[MappingEntryIn]


class SceneBoundary(BaseModel):
    index: int
    start: int
    end: int
    title: str


class AdaptationProjectOut(BaseModel):
    id: int
    title: str
    source_filename: Optional[str]
    intent: Optional[str]
    intensity: int
    era_target: Optional[str]
    status: str
    created_at: datetime
    updated_at: datetime
    word_count: int
    scene_boundaries: List[SceneBoundary] = []
    versions: List["AdaptationVersionOut"] = []
    mappings: List[MappingEntryOut] = []

    model_config = {"from_attributes": True}


class AdaptationVersionOut(BaseModel):
    id: int
    version_no: int
    triggered_by: str
    status: str
    stats: Optional[dict]
    error: Optional[str]
    created_at: datetime
    completed_at: Optional[datetime]

    model_config = {"from_attributes": True}


class SceneResultOut(BaseModel):
    id: int
    scene_index: int
    scene_title: Optional[str]
    status: str
    error: Optional[str]
    token_used: Optional[int]
    line_count_delta_pct: Optional[float]
    original_scene_text: str
    rewritten_scene_text: Optional[str]
    manual_edits: Optional[list] = []
    updated_at: datetime

    model_config = {"from_attributes": True}


class VersionDetailOut(AdaptationVersionOut):
    scene_results: List[SceneResultOut] = []


class RunCreate(BaseModel):
    extra_prompt: Optional[str] = None


class SceneRerunRequest(BaseModel):
    extra_prompt: Optional[str] = None


class SceneManualPatch(BaseModel):
    rewritten_scene_text: str = Field(min_length=0)


class MappingSuggestRequest(BaseModel):
    only_empty: bool = True

File: frontend/src/api/adaptation.ts
import request from './request'
import type { ScoreDocxResponse } from './rubric'

export type EntityType = 'person' | 'place' | 'prop' | 'era_term' | 'other'

export interface MappingEntry {
  id?: number
  entity_type: EntityType
  original_text: string
  replacement_text?: string | null
  locked: boolean
  notes?: string | null
  order_index: number
}

export interface SceneBoundary {
  index: number
  start: number
  end: number
  title: string
}

export interface AdaptationVersion {
  id: number
  version_no: number
  triggered_by: string
  status: 'running' | 'done' | 'partial' | 'failed'
  stats?: any
  error?: string | null
  created_at: string
  completed_at?: string | null
}

export interface SceneResult {
  id: number
  scene_index: number
  scene_title?: string
  status: 'pending' | 'running' | 'done' | 'failed' | 'manual_edited'
  error?: string | null
  token_used?: number
  line_count_delta_pct?: number | null
  original_scene_text: string
  rewritten_scene_text?: string | null
  manual_edits?: any[]
  updated_at: string
}

export interface AdaptationProject {
  id: number
  title: string
  source_filename?: string
  intent?: string | null
  intensity: number
  era_target?: string | null
  status: string
  created_at: string
  updated_at: string
  word_count: number
  scene_boundaries: SceneBoundary[]
  versions: AdaptationVersion[]
  mappings: MappingEntry[]
}

const base = '/adaptation'

export const adaptationApi = {
  list: () => request.get<AdaptationProject[]>(`${base}/projects`),
  get: (id: number) => request.get<AdaptationProject>(`${base}/projects/${id}`),
  createWithText: (payload: {title: string; raw_text: string; intent?: string; intensity: number; era_target?: string}) =>
    request.post<AdaptationProject>(`${base}/projects`, payload),
  createWithUpload: (form: FormData) =>
    request.post<AdaptationProject>(`${base}/projects/upload`, form, {
      headers: {'Content-Type': 'multipart/form-data'},
    }),
  update: (id: number, payload: Partial<{title: string; intent: string; intensity: number; era_target: string}>) =>
    request.patch<AdaptationProject>(`${base}/projects/${id}`, payload),
  remove: (id: number) => request.delete(`${base}/projects/${id}`),
  extract: (id: number) => request.post<AdaptationProject>(`${base}/projects/${id}/extract`),
  split: (id: number) => request.post<AdaptationProject>(`${base}/projects/${id}/split`),
  putMappings: (id: number, entries: MappingEntry[]) =>
    request.put<MappingEntry[]>(`${base}/projects/${id}/mappings`, {entries}),
  suggestMappings: (id: number) =>
    request.post<MappingEntry[]>(`${base}/projects/${id}/mappings/suggest`, {only_empty: true}),
  createRun: (id: number, extra_prompt?: string) =>
    request.post<AdaptationVersion>(`${base}/projects/${id}/runs`, {extra_prompt}),
  listRuns: (id: number) => request.get<AdaptationVersion[]>(`${base}/projects/${id}/runs`),
  getRun: (vid: number) =>
    request.get<AdaptationVersion & {scene_results: SceneResult[]}>(`${base}/runs/${vid}`),
  rerunScene: (vid: number, idx: number, extra_prompt?: string) =>
    request.post<SceneResult>(`${base}/runs/${vid}/scenes/${idx}/rerun`, {extra_prompt}),
  patchScene: (vid: number, idx: number, rewritten: string) =>
    request.patch<SceneResult>(`${base}/runs/${vid}/scenes/${idx}`, {rewritten_scene_text: rewritten}),
  exportUrl: (vid: number, format: 'txt' | 'docx') =>
    `/api/v1/adaptation/runs/${vid}/export?format=${format}`,
  getStreamTicket: (vid: number) =>
    request.post<{ticket: string}>(`${base}/runs/${vid}/stream/ticket`),
  streamUrl: (vid: number, ticket: string) =>
    `/api/v1/adaptation/runs/${vid}/stream?ticket=${encodeURIComponent(ticket)}`,
  /** 用 script_rubric handbook 对当前版本拼接出来的全剧本做即时评分 */
  scoreRun: (vid: number) =>
    request.post<ScoreDocxResponse>(`${base}/runs/${vid}/score`),
}
